### PR TITLE
PMM-10677: JSON packing for telemetry metrics

### PR DESCRIPTION
**Pull request by ritbl**
PMM-10677

Added support of transfomation metric value to the JSON object. More info how it should work you can find here https://confluence.percona.com/pages/viewpage.action?pageId=138486089

FB: https://github.com/Percona-Lab/pmm-submodules/pull/2885

### CODE REVIEW

File: managed/services/telemetry/telemetry_test.go [hunk 1]
```diff
@@ -154,22 +174,15 @@ func TestRunTelemetryService(t *testing.T) {
 				tDistributionMethod: 0,
 				dus:                 tt.fields.dus,
 				portalClient:        tt.mockTelemetrySender(),
+				sendCh:              make(chan *pmmv1.ServerMetric, sendChSize),
 			}
 
-			var wg sync.WaitGroup
```
Comment (ritbl) [inline managed/services/telemetry/telemetry_test.go:159]: don't see why we need that

File: managed/services/telemetry/telemetry_test.go [hunk 2]
```diff
@@ -237,14 +250,14 @@ func getDistributionUtilService(t *testing.T, l *logrus.Entry) *distributionUtil
 	return dus
 }
 
-func initMockTelemetrySender(t *testing.T, expetedReport *reporter.ReportRequest, timesCall int) func() sender {
+func initMockTelemetrySender(t *testing.T, expectedReport *reporter.ReportRequest, timesCall int) func() sender {
```
Comment (ritbl) [inline managed/services/telemetry/telemetry_test.go:253]: typo

File: managed/services/telemetry/deps.go [hunk 3]
```diff
@@ -47,6 +47,6 @@ type DataSourceLocator interface {
 
 // DataSource telemetry data source.
 type DataSource interface {
-	FetchMetrics(ctx context.Context, config Config) ([][]*pmmv1.ServerMetric_Metric, error)
+	FetchMetrics(ctx context.Context, config Config) ([]*pmmv1.ServerMetric_Metric, error)
```
Comment (ritbl) [inline managed/services/telemetry/deps.go:50]: A tiny refactoring to simplify impl

File: managed/services/telemetry/transform.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	firstMetric := config.Data[0].MetricName
+
+	result := map[string]itemsType{}
+
+	var items []map[string]any
+	var next map[string]any
+	for _, metric := range metrics {
+		isFirstMetric := metric.Key == firstMetric
```
Comment (dliakhov) [inline managed/services/telemetry/transform.go:50]: Do you think we need this check now? As I understand now we set metric in datasource's implementations even if metric has empty value.

File: managed/services/telemetry/transform.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+	if len(config.Data) == 0 || config.Data[0].MetricName == "" {
+		return nil, errors.Errorf("invalid metrics config")
+	}
+
+	firstMetric := config.Data[0].MetricName
+
+	result := map[string]itemsType{}
+
+	var items []map[string]any
```
Comment (dliakhov) [inline managed/services/telemetry/transform.go:47]: nit: `var items itemsType`

File: managed/services/telemetry/deps.go (hunk 3 above)
Comment (dliakhov) [reply managed/services/telemetry/deps.go:50]: Nice, love it

File: managed/services/telemetry/transform.go (hunk 4 above)
Comment (dliakhov) [reply managed/services/telemetry/transform.go:50]: Oh, now I got why do we need that.

File: managed/services/telemetry/telemetry.go [hunk 6]
```diff
@@ -233,10 +233,20 @@ func (s *Service) prepareReport(ctx context.Context) *pmmv1.ServerMetric {
...
-		for _, each := range metrics {
-			telemetryMetric.Metrics = append(telemetryMetric.Metrics, each...)
+		if telemetry.Transform != nil {
+			if telemetry.Transform.Type == JSONTransformType {
+				metrics, err = transformToJSON(&telemetry, metrics)
+				if err != nil {
+					s.l.Debugf("failed to transform to JSON: %s", err)
+					continue
+				}
+			}
```
Comment (YaroslavPodorvanov) [inline managed/services/telemetry/telemetry.go:243]: else "undefined transform type %s"?

File: managed/services/telemetry/transform.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	return []*pmmv1.ServerMetric_Metric{
+		{
+			Key:   config.Transform.Metric,
+			Value: string(resultAsJSON),
+		},
+	}, nil
+}
+
+func removeEmpty(metrics []*pmmv1.ServerMetric_Metric) []*pmmv1.ServerMetric_Metric {
+	result := make([]*pmmv1.ServerMetric_Metric, 0, len(metrics))
```
Comment (YaroslavPodorvanov) [inline managed/services/telemetry/transform.go:87]: reuse memory:
```go
result := metrics[:0]
```

File: managed/services/telemetry/telemetry_test.go [hunk 8]
```diff
@@ -39,8 +38,29 @@ import (
...
+)
+
 func TestRunTelemetryService(t *testing.T) {
-	t.Skip()
+	pgHostPort := "127.0.0.1:5432"
+	pgHostPortFromEnv, ok := os.LookupEnv(envPGHostPort)
+	if ok {
+		pgHostPort = pgHostPortFromEnv
+	}
+	qanDSN := "tcp://localhost:9000?database=pmm&block_size=10000&pool_size="
```
Comment (YaroslavPodorvanov) [inline managed/services/telemetry/telemetry_test.go:53]: `block_size` & `pool_size` not available for "github.com/ClickHouse/clickhouse-go/v2"

File: managed/services/telemetry/transform_test.go [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+			name: "happy path",
+			args: args{
+				config: config(),
+				metrics: []*pmmv1.ServerMetric_Metric{
+					{Key: config().Data[0].MetricName, Value: "v1"},
+					{Key: config().Data[0].MetricName, Value: "v2"},
+				},
+			},
+			want: []*pmmv1.ServerMetric_Metric{
+				{Key: config().Transform.Metric, Value: "{\"v\":[{\"my-metric\":\"v1\"},{\"my-metric\":\"v2\"}]}"},
```
Comment (YaroslavPodorvanov) [inline managed/services/telemetry/transform_test.go:102]: just proposal:
```go
			want: []*pmmv1.ServerMetric_Metric{
				{Key: config().Transform.Metric, Value: `"{"v":[{"my-metric":"v1"},{"my-metric":"v2"}]}"`},
			},
```

**Review by YaroslavPodorvanov — APPROVED** [review]

**Review by dliakhov — APPROVED** [review]

File: managed/services/telemetry/transform.go (hunk 7 above)
Comment (dliakhov) [reply managed/services/telemetry/transform.go:87]: are you sure that it's safe? because in this case we will override existing slice in the loop.

```	
result := metrics[:0]
for _, metric := range metrics {
  if metric.Value != "" {
	result = append(result, metric)
  }
}
```

File: managed/services/telemetry/transform_test.go (hunk 9 above)
Comment (dliakhov) [reply managed/services/telemetry/transform_test.go]: fixed.

File: managed/services/telemetry/telemetry_test.go (hunk 8 above)
Comment (dliakhov) [reply managed/services/telemetry/telemetry_test.go]: fixed.

File: managed/services/telemetry/telemetry.go (hunk 6 above)
Comment (dliakhov) [reply managed/services/telemetry/telemetry.go:245]: good catch

fixed

**Review by gen1us2k — APPROVED** [review]

File: managed/services/telemetry/telemetry.go (hunk 6 above)
Comment (ritbl) [reply managed/services/telemetry/telemetry.go:246]: I'd reword `undefined`. `unsupported` is a more correct word here

File: managed/services/telemetry/transform.go (hunk 4 above)
Comment (ritbl) [reply managed/services/telemetry/transform.go:51]: added comment to make it a bit more clear

File: managed/services/telemetry/transform.go (hunk 7 above)
Comment (ritbl) [reply managed/services/telemetry/transform.go:88]: Reusing memory is almost never a good idea
